Library/Homebrew/language/python.rb

```diff
@@ -96,5 +97,83 @@ def self.setup_install_args(prefix)
     def self.package_available?(python, module_name)
       quiet_system python, "-c", "import #{module_name}"
     end
+
+    def self.virtualenv(formula, venv_root, python="python")
+      venv = Virtualenv.new venv_root, python
+      # Do some cute metaprogramming so that when `system` is invoked by the
+      # Virtualenv instance, `Formula#system` is called, so that we get the
+      # output capture and logging behavior we expect from formulae.
+      metaclass = class << venv; self; end
+      metaclass.send :define_method, :system do |*args|
+        formula.system(*args)
+      end
+      venv.create!
+      yield venv
+    end
+
+    class Virtualenv
+      def initialize(venv_root, python)
+        @venv_root = Pathname.new(venv_root)
+        @python = python
+      end
+
+      def create!
+        virtualenv_resource = Resource.new "virtualenv"
+        virtualenv_resource.url "https://files.pythonhosted.org/packages/5c/79/5dae7494b9f5ed061cff9a8ab8d6e1f02db352f3facf907d9eb614fb80e9/virtualenv-15.0.2.tar.gz"
+        virtualenv_resource.sha256 "fab40f32d9ad298fba04a260f3073505a16d52539a84843cf8c8369d4fd17167"
```

Comment:
Thanks for the review, btw. Can you explain how the `module` technique would be different from the current implementation? I don't understand the distinction, sorry.

Why do you prefer making virtualenv a resource on the Formula class instead of instantiating a Resource directly?